File: VB/HideableGroupRowFooters/CustomGridControl/MyGridViewInfo.vb
Imports System.Drawing
Imports DevExpress.XtraGrid.Views.Grid
Imports DevExpress.XtraGrid.Views.Grid.ViewInfo
Imports System.Diagnostics

Namespace HideableGroupRowFooters

	Public Class MyGridViewInfo
		Inherits GridViewInfo

		Public Sub New(ByVal gridView As MyGridView)
			MyBase.New(gridView)
		End Sub

		Protected Overrides Sub CalcRowFooterInfo(ByVal ri As GridRowInfo, ByVal row As GridRow, ByVal nextRow As GridRow)
			Dim height As Integer = ri.RowFooters.RowFootersHeight
			If height = 0 Then
				Return
			End If

			Dim isShowCurrentFooter As Boolean = IsShowCurrentRowFooter(ri)
			Dim startLevel As Integer = ri.Level
			Dim footerRowHandle As Integer = ri.RowHandle

			If (Not ri.IsGroupRow) OrElse (Not isShowCurrentFooter) Then
				footerRowHandle = View.GetParentRowHandle(footerRowHandle)
			End If

			If Not isShowCurrentFooter Then
				startLevel -= 1
			End If

			Dim top As Integer = ri.TotalBounds.Bottom - height - ri.RowSeparatorBounds.Height
			Dim left As Integer = ri.IndentRect.Right - (If((Not isShowCurrentFooter), LevelIndent, 0))
			ri.RowFooters.Bounds = New Rectangle(left, top, ri.DataBounds.Right - left, height)

			Dim n As Integer = 0
			Do While n < ri.RowFooters.RowFooterCount

				Dim some As Boolean = NeedHideFooter(row.VisibleIndex+1, startLevel)

				If Not some Then
					startLevel -= 1
					left -= LevelIndent
					ri.RowFooters.RowFooterCount += 1
					footerRowHandle = View.GetParentRowHandle(footerRowHandle)

					n += 1
					Continue Do
				End If

				Dim fi As New GridRowFooterInfo()
				ri.RowFooters.Add(fi)
				fi.RowHandle = footerRowHandle
				fi.Bounds = ri.Bounds
				fi.Level = startLevel
				fi.Bounds.Y = top
				fi.Bounds.X = left
				fi.Bounds.Width = ri.DataBounds.Right - fi.Bounds.Left
				fi.Bounds.Height = GroupFooterHeight
				top += fi.Bounds.Height

				If Not ri.IndicatorRect.IsEmpty Then
					fi.IndicatorRect = ri.IndicatorRect
					fi.IndicatorRect.Y = fi.Bounds.Y
					fi.IndicatorRect.Height = fi.Bounds.Height
				End If

				If View.OptionsView.ShowHorizontalLines <> DevExpress.Utils.DefaultBoolean.False Then
					ri.AddRowLineInfo(fi.Bounds.Left, fi.Bounds.Bottom - 1, fi.Bounds.Width, 1, PaintAppearance.HorzLine)
					fi.Bounds.Height -= 1
				End If

				CalcRowCellsFooterInfo(fi, ri)
				footerRowHandle = View.GetParentRowHandle(footerRowHandle)
				startLevel -= 1
				left -= LevelIndent
				n += 1
			Loop
		End Sub

		Public Overrides Function GetRowFooterCount(ByVal rowHandle As Integer, ByVal rowVisibleIndex As Integer, ByVal isExpanded As Boolean) As Integer
			Dim initialVisibleFootersCount As Integer = MyBase.GetRowFooterCount(rowHandle, rowVisibleIndex, isExpanded)
			Dim visibleFootersCount As Integer = initialVisibleFootersCount

			Dim footerRowHandle As Integer = rowHandle
			For i As Integer = 0 To initialVisibleFootersCount - 1
				Dim some As Boolean = NeedHideFooter(rowVisibleIndex+1, View.GetRowLevel(footerRowHandle))

				If Not some Then
					visibleFootersCount -= 1
				End If

				footerRowHandle = View.GetParentRowHandle(footerRowHandle)
			Next i

			Return visibleFootersCount
		End Function
		Public Function NeedHideFooter(ByVal rowHandle As Integer, ByVal footerLevel As Integer) As Boolean
			CurentFooterLevel = 0
			CalcFooterLevels(rowHandle)
			If rowHandle <> Integer.MinValue + 1 AndAlso rowHandle<0 Then
				Return (HidingGroupFooters(rowHandle, footerLevel))
			End If


			CurentFooterLevel = 0
			LastGroupRowHandle = 0
			GetLastGroupRow()
			CalcFooterLevels(LastGroupRowHandle)
			Return (HidingGroupFooters(LastGroupRowHandle, footerLevel))

		End Function

		Private Sub RaiseHideGroupFooter(ByVal args As HideFooterEventArgs)
            Dim aView As MyGridView = TryCast(View, MyGridView)
            If aView IsNot Nothing Then
                aView.OnCustomFooterVisible(args)
            End If

		End Sub


		Private CurentFooterLevel As Integer
		Private LastGroupRowHandle As Integer
		Public Sub GetLastGroupRow()
			Dim k As Integer = 0
			Do
				k -= 1
				If View.IsValidRowHandle(k) Then
					If View.IsGroupRow(k) Then
						LastGroupRowHandle = k
					End If
				End If


			Loop While View.IsValidRowHandle(k)
		End Sub


		Public Function HidingGroupFooters(ByVal rowHandle As Integer, ByVal footerLevel As Integer) As Boolean

			If rowHandle = -2147483648 Then
				Return True
			End If

			Dim args As New HideFooterEventArgs(View.GetChildRowCount(rowHandle))
			RaiseHideGroupFooter(args)
			If CurentFooterLevel = footerLevel Then
				Return args.FooterVisible
			End If
			Dim n As Integer = View.GetParentRowHandle(rowHandle)
			If n < 0 AndAlso n <> -2147473648 Then
				CurentFooterLevel -= 1
			  Return HidingGroupFooters(n,footerLevel)
			End If
			Return True

		End Function

		Public Sub CalcFooterLevels(ByVal rowHandle As Integer)
			CurentFooterLevel += 1
			Dim n As Integer = View.GetParentRowHandle(rowHandle)
			If n < 0 AndAlso n <> -2147483648 Then
				CalcFooterLevels(n)
			End If

		End Sub
	End Class
End Namespace

File: CS/HideableGroupRowFooters/CustomGridControl/MyGridViewInfo.cs
using System.Drawing;
using DevExpress.XtraGrid.Views.Grid;
using DevExpress.XtraGrid.Views.Grid.ViewInfo;
using System.Diagnostics;

namespace HideableGroupRowFooters
{

	public class MyGridViewInfo : GridViewInfo
	{
		public MyGridViewInfo(MyGridView gridView)
			: base(gridView)
		{
		}

        protected override void CalcRowFooterInfo(GridRowInfo ri, GridRow row, GridRow nextRow)
        {
            int height = ri.RowFooters.RowFootersHeight;
            if (height == 0)
                return;

            bool isShowCurrentFooter = IsShowCurrentRowFooter(ri);
            int startLevel = ri.Level;
            int footerRowHandle = ri.RowHandle;

            if (!ri.IsGroupRow || !isShowCurrentFooter)
                footerRowHandle = View.GetParentRowHandle(footerRowHandle);

            if (!isShowCurrentFooter)
            {
                startLevel--;
            }

            int top = ri.TotalBounds.Bottom - height - ri.RowSeparatorBounds.Height;
            int left = ri.IndentRect.Right - (!isShowCurrentFooter ? LevelIndent : 0);
            ri.RowFooters.Bounds = new Rectangle(left, top, ri.DataBounds.Right - left, height);

            for (int n = 0; n < ri.RowFooters.RowFooterCount; n++)
            {

                bool some = NeedHideFooter(row.VisibleIndex+1, startLevel);

                if (!some)
                {
                    startLevel -= 1;
                    left -= LevelIndent;
                    ri.RowFooters.RowFooterCount += 1;
                    footerRowHandle = View.GetParentRowHandle(footerRowHandle);

                    continue;
                }

                GridRowFooterInfo fi = new GridRowFooterInfo();
                ri.RowFooters.Add(fi);
                fi.RowHandle = footerRowHandle;
                fi.Bounds = ri.Bounds;
                fi.Level = startLevel;
                fi.Bounds.Y = top;
                fi.Bounds.X = left;
                fi.Bounds.Width = ri.DataBounds.Right - fi.Bounds.Left;
                fi.Bounds.Height = GroupFooterHeight;
                top += fi.Bounds.Height;

                if (!ri.IndicatorRect.IsEmpty)
                {
                    fi.IndicatorRect = ri.IndicatorRect;
                    fi.IndicatorRect.Y = fi.Bounds.Y;
                    fi.IndicatorRect.Height = fi.Bounds.Height;
                }

                if (View.OptionsView.ShowHorizontalLines != DevExpress.Utils.DefaultBoolean.False)
                {
                    ri.AddRowLineInfo(fi.Bounds.Left, fi.Bounds.Bottom - 1,
                        fi.Bounds.Width, 1, PaintAppearance.HorzLine);
                    fi.Bounds.Height -= 1;
                }

                CalcRowCellsFooterInfo(fi, ri);
                footerRowHandle = View.GetParentRowHandle(footerRowHandle);
                startLevel--;
                left -= LevelIndent;
            }
        }

        public override int GetRowFooterCount(int rowHandle, int rowVisibleIndex, bool isExpanded)
        {
            int initialVisibleFootersCount = base.GetRowFooterCount(rowHandle, rowVisibleIndex, isExpanded);
            int visibleFootersCount = initialVisibleFootersCount;

            int footerRowHandle = rowHandle;
            for (int i = 0; i < initialVisibleFootersCount; i++)
            {
                bool some = NeedHideFooter(rowVisibleIndex+1, View.GetRowLevel(footerRowHandle));

                if (!some)
                    visibleFootersCount--;

                footerRowHandle = View.GetParentRowHandle(footerRowHandle);
            }

            return visibleFootersCount;
        }
        public bool NeedHideFooter(int rowHandle,int footerLevel)
        {
            CurentFooterLevel = 0;
            CalcFooterLevels(rowHandle);
            if (rowHandle != int.MinValue + 1 && rowHandle<0)
                return (HidingGroupFooters(rowHandle, footerLevel));


            CurentFooterLevel = 0;
            LastGroupRowHandle = 0;
            GetLastGroupRow();
            CalcFooterLevels(LastGroupRowHandle);
            return (HidingGroupFooters(LastGroupRowHandle, footerLevel));

        }

        private void RaiseHideGroupFooter(HideFooterEventArgs args)
        {
            MyGridView view = View as MyGridView;
            if (view != null)
                view.OnCustomFooterVisible(args);
        
        }


        int CurentFooterLevel;
        int LastGroupRowHandle;
        public void GetLastGroupRow()
        {
            int k = 0;
            do
            {
                k--;
                if (View.IsValidRowHandle(k))
                    if (View.IsGroupRow(k))
                        LastGroupRowHandle = k;


            }
            while (View.IsValidRowHandle(k));
        }


        public bool HidingGroupFooters(int rowHandle,int footerLevel)
        {

            if (rowHandle == -2147483648)
                return true;

            HideFooterEventArgs args = new HideFooterEventArgs(View.GetChildRowCount(rowHandle));
            RaiseHideGroupFooter(args);
            if (CurentFooterLevel == footerLevel)
            {
                return args.FooterVisible;
            }
            int n = View.GetParentRowHandle(rowHandle);
            if (n < 0 && n != -2147473648)
            {
                CurentFooterLevel--;
              return HidingGroupFooters(n,footerLevel);
            }
            return true;

        }

        public void CalcFooterLevels(int rowHandle)
        {
            CurentFooterLevel++;
            int n = View.GetParentRowHandle(rowHandle);
            if (n < 0 && n != -2147483648)
                CalcFooterLevels(n);

        }
	}
}
